CSharp/Library/Microsoft.Bot.Connector.Shared/ActivityEx.cs

```diff
@@ -174,28 +212,82 @@ public StateClient GetStateClient(MicrosoftAppCredentials credentials, string se
         /// <param name="microsoftAppPassword"></param>
         /// <param name="serviceUrl">alternate serviceurl to use for state service</param>
         /// <param name="handlers"></param>
+        /// <param name="activity"></param>
         /// <returns></returns>
-        public StateClient GetStateClient(string microsoftAppId = null, string microsoftAppPassword = null, string serviceUrl = null, params DelegatingHandler[] handlers)
+        public static StateClient GetStateClient(this IActivity activity, string microsoftAppId = null, string microsoftAppPassword = null, string serviceUrl = null, params DelegatingHandler[] handlers) => GetStateClient(activity, new MicrosoftAppCredentials(microsoftAppId, microsoftAppPassword), serviceUrl, handlers);
+
+        /// <summary>
+        /// Return the "major" portion of the activity
+        /// </summary>
+        /// <param name="activity"></param>
+        /// <returns>normalized major portion of the activity, aka message/... will return "message"</returns>
+        public static string GetActivityType(this IActivity activity)
+        {
+            var type = activity.Type.Split('/').First();
+            return Activity.GetActivityType(type);
+        }
+
+        /// <summary>
+        /// Get channeldata as typed structure
+        /// </summary>
+        /// <param name="activity"></param>
+        /// <typeparam name="TypeT">type to use</typeparam>
+        /// <returns>typed object or default(TypeT)</returns>
+        public static TypeT GetChannelData<TypeT>(this IActivity activity)
         {
-            return GetStateClient(new MicrosoftAppCredentials(microsoftAppId, microsoftAppPassword), serviceUrl, handlers);
+            if (activity.ChannelData == null)
+                return default(TypeT);
+            return ((JObject)activity.ChannelData).ToObject<TypeT>();
         }
 
+
+        /// <summary>
+        /// Get channeldata as typed structure
+        /// </summary>
+        /// <param name="activity"></param>
+        /// <typeparam name="TypeT">type to use</typeparam>
+        /// <param name="instance">The resulting instance, if possible</param>
+        /// <returns>
+        /// <c>true</c> if value of <seealso cref="IActivity.ChannelData"/> was coerceable to <typeparamref name="TypeT"/>, <c>false</c> otherwise.
+        /// </returns>
+        public static bool TryGetChannelData<TypeT>(this IActivity activity,
+            out TypeT instance)
+        {
+            instance = default(TypeT);
+
+            try
+            {
+                if (activity.ChannelData == null)
+                {
+                    return false;
+                }
+
+                instance = GetChannelData<TypeT>(activity);
+                return true;
+            }
+            catch

```

Comment:
What would be the appropriate way to achieve this behavior without duplicating the code w/in `GetChannelData`?